docs/en/sql-reference/functions/nlp-functions.md

```diff
@@ -0,0 +1,155 @@
+---
+toc_priority: 67
+toc_title: NLP
+---
+
+# Natural Language Processing functions {#nlp-functions}
+
+## tokenize, tokenizeWhitespace {#tokenize}
+
+Performs tokenization on a given text.
+
+**Syntax**
+
+``` sql
+tokenize(text)
+```
+
+**Arguments**
+
+-   `text` — Text that needs to be tokenized. [String](../../sql-reference/data-types/string.md#string).
+
+**Examples**
+
+Query:
+
+``` sql
+SELECT tokenize('I think it is a blessing in disguise.') as res;
+```
+
+Result:
+
+``` text
+┌─res────────────────────────────────────────────────────┐
+│ ['I','think','it','is','a','blessing','in','disguise'] │
+└────────────────────────────────────────────────────────┘
+```
+
+## stem {#stem}
+
+Performs stemming on a previously tokenized text.
+
+**Syntax**
+
+``` sql
+stem('language', word)
+```
+
+**Arguments**
+
+-   `language` — Language which rules will be applied. Must be in lowercase. [String](../../sql-reference/data-types/string.md#string).
+-   `word` — word that needs to be stemmed. Must be in lowercase. [String](../../sql-reference/data-types/string.md#string).
+
+**Examples**
+
+Query:
+
+``` sql
+SELECT SELECT arrayMap(x -> stem('en', x), ['I', 'think', 'it', 'is', 'a', 'blessing', 'in', 'disguise']) as res;
+```
+
+Result:
+
+``` text
+┌─res────────────────────────────────────────────────┐
+│ ['I','think','it','is','a','bless','in','disguis'] │
+└────────────────────────────────────────────────────┘
+```
+
+## lemmatize {#lemmatize}
+
+Performs lemmatization on a given word.
+
+**Syntax**
+
+``` sql
+lemmatize('language', word)
+```
+
+**Arguments**
+
+-   `language` — Language which rules will be applied. [String](../../sql-reference/data-types/string.md#string).
+-   `word` — Word that needs to be lemmatized. Must be lowercase. [String](../../sql-reference/data-types/string.md#string).
+
+**Examples**
+
+Query:
+
+``` sql
+SELECT lemmatize('en', 'wolves');
+```
+
+Result:
+
+``` text
+┌─lemmatize("wolves")─┐
+│              "wolf" │
+└─────────────────────┘
+```
+
+Configuration:
+``` xml
+<lemmatizers>
+    <lemmatizer>
+        <lang>en</lang>
+        <path>en.bin</path>
```

Comment:
It's not clear which format of this file should be used. Any references? Also why we don't embed some dictionaries by default? At least for the English language.

Reply:
It uses pretrained LemmaGen dictionaries.

Reply:
It uses pretrained LemmaGen dictionaries.

There are 2 main sources of them that I have found (and both of them work):
1. From [this](https://pypi.org/project/Lemmagen/) package from PyPi (whence my implementation comes from): https://github.com/evillique/lemmagen-c/tree/main/data
2. Additional dictionaries (which seem better to me and include more languages including Russian): https://github.com/vpodpecan/lemmagen3/tree/master/src/lemmagen3/models

Speaking about the licence, did not found any information about (1) pretrained dictionaries, and (2) says [that](https://github.com/vpodpecan/lemmagen3#license) about it:
 
![image](https://user-images.githubusercontent.com/43110995/122656715-e2afb480-d165-11eb-8b4d-a4284d03ff91.png)

It's probably worth to clarify legal issues before embedding any dictionaries.